torch/_dynamo/output_graph.py

```diff
@@ -441,24 +470,51 @@ def wrap_name(module_key):
-                return wrap_name(k)
+        # Each param has a source, and even if they have the same tensor value, it needs to
+        # be reachable. So, for ParamBufferSource, which is the leaf node source initialized below
+        # we can skip this value check. It's wrap_name is a no op anyway, so this is safe.
+        if not isinstance(source, ParamBufferSource):
+            for k, v in self.nn_modules.items():
+                if v is target:
+                    # it already exists, but register it anyway
+                    return wrap_name(k)
 
         # create a new unique name
         name = "_".join(map(str, names))
-        # e.g. replace abc.xyz[123].qkv with abc.xyz_123.qkv
-        name = re.sub(r"\[(\d+)\]", r"_\g<1>", name)
-        # e.g. replace abc.xyz_123.qkv with abc_xyz_123_qkv
-        name = re.sub(r"[^a-zA-Z0-9]", "_", name)
 
+        # NOTE! This code used to strip square brackets,
+        # but we no longer do so in favor of a single uniformly used
+        # normalize_attr_name. We used to need this because of things we did with module
+        # lookups where we would eval with a name, and the square brackets would throw it off.
+        # We no longer do this.
+        #
+        # The reason we need this uniform is that we need all the names to be the same, so that
+        # all the sources can be the same, because when we eventually call .register into the
+        # context_module, we need to make sure elements match.
+        name = normalize_attr_name(name)
         if not name or not name[0].isalpha():
             name = "sub" + name
         base = name
         for i in itertools.count():
             if name not in self.nn_modules:
                 self.nn_modules[name] = target
+                assert self.names_to_sources is not None
+                self.tracing_context.module_context.register(source)
+                if isinstance(target, torch.nn.Module) and not is_lazy_module(target):
+                    # annoying, but there are cases when we do not have parameters
+                    # see test_nn_moduledict_contains
+                    if hasattr(target, "_parameters"):
+                        for n, p in target.named_parameters(recurse=False, remove_duplicate=False):
+                            new_source = ParamBufferSource(source, n)
+                            new_name = new_source.name()
+                            self.register_attr_or_module(p, new_name, source=new_source)
+                    # annoying, but there are cases when we do not have buffers
+                    # see test_nn_moduledict_contains
+                    if hasattr(target, "_buffers"):
+                        for n, p in target.named_buffers(recurse=False, remove_duplicate=False):
+                            new_source = ParamBufferSource(source, n)
+                            new_name = new_source.name()
+                            self.register_attr_or_module(p, new_name, source=new_source)
```

Comment:
Reviewers - do you prefer we share the 3 common lines here, or leave it like this for readability?